### ISSUE TITLE
flang: falure compiling blas sources

### ISSUE THREAD

**Issue by balay** [llvm/llvm-project#62551]
```console
$ flang-new -o cbdsqr.o -c -fPIC -O0 cbdsqr.f
<snip>
/usr/bin/ranlib libfblas.a/usr/include/c++/13/optional:484: constexpr const _Tp& std::_Optional_base_impl<_Tp, _Dp>::_M_get() const [with _Tp = long unsigned int; _Dp = std::_Optional_base<long unsigned int, true, true>]: Assertion 'this->_M_is_engaged()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /usr/bin/flang-new -fc1 -triple x86_64-redhat-linux-gnu -emit-obj -mrelocation-model pic -pic-level 2 -target-cpu x86-64 -O0 -o cbdsqr.o -x f95-cpp-input cbdsqr.f
 #0 0x00007fb4302c57aa llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM-16.so+0xcc57aa)
 #1 0x00007fb4302c3094 llvm::sys::RunSignalHandlers() (/lib64/libLLVM-16.so+0xcc3094)
 #2 0x00007fb4302c635b (/lib64/libLLVM-16.so+0xcc635b)
 #3 0x00007fb42f05fb70 __restore_rt (/lib64/libc.so.6+0x3db70)
 #4 0x00007fb42f0b0844 __pthread_kill_implementation (/lib64/libc.so.6+0x8e844)
 #5 0x00007fb42f05fabe gsignal (/lib64/libc.so.6+0x3dabe)
 #6 0x00007fb42f04887f abort (/lib64/libc.so.6+0x2687f)
 #7 0x00007fb42f2df1c0 __gnu_debug::_Safe_sequence_base::_M_get_mutex() (/lib64/libstdc++.so.6+0xdf1c0)
 #8 0x00007fb42eabf100 (/lib64/libFortranLower.so.16+0x4bf100)
 #9 0x00007fb42eac1728 (/lib64/libFortranLower.so.16+0x4c1728)
#10 0x00007fb42eace252 (/lib64/libFortranLower.so.16+0x4ce252)
#11 0x00007fb42eacd9bd (/lib64/libFortranLower.so.16+0x4cd9bd)
#12 0x00007fb42ead4bce (/lib64/libFortranLower.so.16+0x4d4bce)
#13 0x00007fb42ead5174 Fortran::lower::createPFT(Fortran::parser::Program const&, Fortran::semantics::SemanticsContext const&) (/lib64/libFortranLower.so.16+0x4d5174)
#14 0x00007fb42e7c7b15 Fortran::lower::LoweringBridge::lower(Fortran::parser::Program const&, Fortran::semantics::SemanticsContext const&) (/lib64/libFortranLower.so.16+0x1c7b15)
#15 0x00007fb43aeee8b2 Fortran::frontend::CodeGenAction::beginSourceFileAction() (/lib64/libflangFrontend.so.16+0xa68b2)
#16 0x00007fb43aede665 Fortran::frontend::FrontendAction::beginSourceFile(Fortran::frontend::CompilerInstance&, Fortran::frontend::FrontendInputFile const&) (/lib64/libflangFrontend.so.16+0x96665)
#17 0x00007fb43aece29d Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (/lib64/libflangFrontend.so.16+0x8629d)
#18 0x00007fb43b0a38eb Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (/lib64/libflangFrontendTool.so.16+0x6e8eb)
#19 0x000055faad9fc0a8 fc1_main(llvm::ArrayRef<char const*>, char const*) (/usr/bin/flang-new+0x80a8)
#20 0x000055faad9f9a64 main (/usr/bin/flang-new+0x5a64)
#21 0x00007fb42f049b4a __libc_start_call_main (/lib64/libc.so.6+0x27b4a)
#22 0x00007fb42f049c0b __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c0b)
#23 0x000055faad9f9c85 _start (/usr/bin/flang-new+0x5c85)
flang-new: error: unable to execute command: Aborted (core dumped)
flang-new: error: flang frontend command failed due to signal (use -v to see invocation)
flang-new version 16.0.1 (Fedora 16.0.1-1.fc38)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
flang-new: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
flang-new: note: diagnostic msg: /tmp/cbdsqr-ba5cfc
flang-new: note: diagnostic msg: /tmp/cbdsqr-ba5cfc.sh
flang-new: note: diagnostic msg: 

********************
```
[cbdsqr-ba5cfc.sh.txt](https://github.com/llvm/llvm-project/files/11399341/cbdsqr-ba5cfc.sh.txt)
[cbdsqr-ba5cfc.txt](https://github.com/llvm/llvm-project/files/11399342/cbdsqr-ba5cfc.txt)

**Comment by balay** [comment]
> Which version are you using

flang-new version 16.0.1 (Fedora 16.0.1-1.fc38)

> Can you try with flang upstream?

You mean latest 16 or 17? 

Could you point me to instructions on how I can install the latest upstream? I might just have to wait for fedora to package it..

**Comment by balay** [comment]
I attempted to build flang via spack. But I can't reproduce this failure . i.e fails with fedora flang 16.0.2, but works with spack installed flang 16.0.2 (and 16.0.3, main branch)
```
[balay@pj01 lapack]$ /home/balay/spack/opt/spack/linux-fedora38-skylake/gcc-13.0.1/llvm-16.0.2-mabiq7qo452jrzqiswcdi7ha5w7ap2id/bin/flang-new --version
flang-new version 16.0.2 (https://github.com/spack/spack 16613408e4047542c61322ca7c165fea695f6d4c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/balay/spack/opt/spack/linux-fedora38-skylake/gcc-13.0.1/llvm-16.0.2-mabiq7qo452jrzqiswcdi7ha5w7ap2id/bin
[balay@pj01 lapack]$ /home/balay/spack/opt/spack/linux-fedora38-skylake/gcc-13.0.1/llvm-16.0.2-mabiq7qo452jrzqiswcdi7ha5w7ap2id/bin/flang-new -o cbdsqr.o -c -fPIC -O0 cbdsqr.f
[balay@pj01 lapack]$ echo $?
0
[balay@pj01 lapack]$ /usr/bin/flang-new --version
flang-new version 16.0.2 (Fedora 16.0.2-1.fc38)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
[balay@pj01 lapack]$ /usr/bin/flang-new -o cbdsqr.o -c -fPIC -O0 cbdsqr.f |& head -2
/usr/include/c++/13/optional:484: constexpr const _Tp& std::_Optional_base_impl<_Tp, _Dp>::_M_get() const [with _Tp = long unsigned int; _Dp = std::_Optional_base<long unsigned int, true, true>]: Assertion 'this->_M_is_engaged()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
[balay@pj01 lapack]$ 
```
Perhaps there is a difference in build options here - resulting in the difference in behavior

**Comment by clementval** [comment]
Spack might fetch the latest source to build flang. Can you tell me which spack package you used to build with spack? Anyway it's likely that the backport PR (https://github.com/llvm/llvm-project-release-prs/pull/429) will be merged into 16.0.4.

**Comment by balay** [comment]
Spack builds the specified version (from release tarballs) - I built 3 versions: 16.0.2,  16.0.3, main-branch

https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm/package.py

I can recheck 16.0.4 - when it lands in fedora

[and I'm seeing errors with missing features - so will have to wait on them anyway]

**Comment by clementval** [comment]
Things are moving fast so I suggest to check with the main-branch from the time being.

**Comment by balay** [comment]
> I can recheck 16.0.4 - when it lands in fedora

Closing this issue - as flang-17 in fedora does not give this error.
```
[balay@pj01 lapack]$ /usr/bin/flang-new --version
flang-new version 17.0.6 (Fedora 17.0.6-1.fc39)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
[balay@pj01 lapack]$ /usr/bin/flang-new -o cbdsqr.o -c -fPIC -O0 cbdsqr.f
[balay@pj01 lapack]$ echo $?
0
[balay@pj01 lapack]$ 
```
